File: 1. Creational Design Patterns/Builder/C#/BuilderTemplate.cs
// Builder Pattern – C# Template
using System;

namespace BuilderTemplate
{
    // Product
    class Product { /* ... */ }

    // Builder Interface
    interface IBuilder
    {
        void BuildPartA();
        void BuildPartB();
        Product GetResult();
    }

    // Concrete Builder
    class ConcreteBuilder : IBuilder
    {
        private readonly Product _product = new Product();
        public void BuildPartA() { /* ... */ }
        public void BuildPartB() { /* ... */ }
        public Product GetResult() => _product;
    }

    // Director
    class Director
    {
        public void Construct(IBuilder builder)
        {
            builder.BuildPartA();
            builder.BuildPartB();
        }
    }

    class Program
    {
        static void Main()
        {
            var builder = new ConcreteBuilder();
            new Director().Construct(builder);
            var product = builder.GetResult();
        }
    }
}

File: 1. Creational Design Patterns/Builder/Java/BuilderTemplate.java
// Builder Pattern – Java Template
interface Builder {
    void buildPartA();
    void buildPartB();
    Product getResult();
}
class Product {}
class ConcreteBuilder implements Builder {
    private final Product product = new Product();
    public void buildPartA(){}
    public void buildPartB(){}
    public Product getResult(){return product;}
}
class Director {
    void construct(Builder b){
        b.buildPartA();
        b.buildPartB();
    }
}
public class BuilderTemplate {
    public static void main(String[] args){
        Builder builder = new ConcreteBuilder();
        new Director().construct(builder);
    }
}
